File: basic/algorithm/프로그래머스/DFS/BFS/타겟넘버.js
let answer = 0;

function DFS(arr, target, sum) {
  if (arr.length === 0) {
    if (target === sum) {
      answer += 1;
    }
    return;
  }
  DFS(arr.slice(1), target, sum + arr[0]);
  DFS(arr.slice(1), target, sum - arr[0]);
}

function solution(numbers, target) {
  DFS(numbers, target, 0);
  return answer;
}

console.log(solution([1, 1, 1, 1, 1], 3));

File: basic/algorithm/프로그래머스/DFS/타겟넘버.py
answer = 0

def DFS(numbers, target, result):
    global answer
    if numbers ==[]: 
        if target == result:
            answer +=1
        return
    
    DFS(numbers[1:], target, result+numbers[0])
    DFS(numbers[1:], target, result-numbers[0])

def solution(numbers, target):
    global answer
    DFS(numbers, target, 0)
    return answer